### PR TITLE
Fix jinja block parsing

### LINKED ISSUE #1395: RPC error using `set`
## Issue
The jinja block parser does not know how to handle blocks that have no `end` tag.

`set` is obvious, `do` is also a problem. We should make sure there aren't any others.

### Issue description
In general terms, please describe the issue. What command did you run?

Running commands against `dbt rpc` that have a top-level `set` that does not use the full-block form triggers an error.

### Results
What happened? What did you expect to happen?
Expected: A result
Got: An error

### System information
dbt version: `dev/wilt-chamberlain` (4dd80567e15)
The operating system you're running on: macOS

The python version you're using (probably the output of `python --version`):
python 3.6.5

### Steps to reproduce
1) Run the rpc server with `dbt rpc`
2) Make a query that uses `{% set x = 1 %}` or similar (not a full `set` block), such as `'{% set x = 1 %}select {{ x }} as id'`
3) Get an error like `'Compilation Error\n  unexpected EOF, expected "endset", got "select {{ x }} as id"'`

sample request:
```
{
  'jsonrpc': '2.0',
  'method': 'run',
  'params': {
    'sql': 'eyUgc2V0IHggPSAxICV9c2VsZWN0IHt7IHggfX0gYXMgaWQ=',
    'name': 'foo',
  },
  'id': 1
}
```

That `sql` is just `base64.encodebytes(b'{% set x = 1 %}select {{ x }} as id')`

### PR DESCRIPTION
**Pull request by beckjake**
Fixes #1395

This PR fixes a number of bugs around parsing jinja blocks:

- After we find the start of a comment block, advance immediately
     - this is so we do not mistake "{#}" as both open and close of comment, but instead `}` is the first character of the comment.
     - I couldn't find the issue for this one, if it exists
- support do/set statements (no `{% enddo %}`/`{% endset %}`)
- fix some edge-case bugs around quoting
    - discovered during testing do/set
- fix a bug around materialization parsing
    - discovered during testing do/set
- more tests (including one really nasty one)

### CODE REVIEW

File: core/dbt/clients/_jinja_blocks.py [hunk 1]
```diff
@@ -251,21 +301,36 @@ def find_block(self):
...
+        if block_type_name == 'materialization':
             self.advance(match.end())
             self.handle_materialization(match)
             return True
 
+        if block_type_name == 'do':
+            # if there is a "block_name" in the match groups, we don't expect a
+            # block as the "block name" is actually part of the do-statement.
+            expect_block = matchgroups.get('block_name') is None
+            self.handle_do(match, expect_block=expect_block)
```
Comment (cmcarthur) [inline core/dbt/clients/_jinja_blocks.py]: seems like you shouldn't need to pass `expect_block` along here since `handle_do` determines whether this is a block or a single tag

Comment (beckjake) [reply core/dbt/clients/_jinja_blocks.py]: Oops, good catch - I lost that in one of the many refactors during the development of this PR!

Comment (beckjake) [reply core/dbt/clients/_jinja_blocks.py]: Oh wait, no - we need that for the (weird!) case of: `{% do thing %}`

We could remove this by changing the regex code to understand do-tags explicitly, or by changing `handle_do` to backtrack to the last whitespace if there is a `block_name` in the matchgroups, but I don't think either of those ideas are as good as just doing it this way.

I've added a corresponding test case and a more explanatory comment.

Comment (cmcarthur) [reply core/dbt/clients/_jinja_blocks.py]: hmm.... ok, makes sense.

Comment (beckjake) [reply core/dbt/clients/_jinja_blocks.py]: One way to think about this: if we don't see a "block name", it still might be a statement - for instance `{% do (x = 1234 + my_value) %}` will fail the regex capture for `block_name` and we won't expect a block from this regex, but then `handle_do` will see that there's more data after the `do` and before the `%}` and react appropriately. But if you do see a "block name", you _must_ be inside a statement and therefore should not expect an `{% enddo %}` tag.
